File: src/Java/213.HouseRobberII.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1)
            return nums[0];
        int[] dpa = new int[n + 1];
        int[] dpb = new int[n + 1];
        for (int i = 2; i < n + 1; i++) {
            dpa[i] = Math.max(dpa[i - 1], dpa[i - 2] + nums[i - 2]);
            dpb[i] = Math.max(dpb[i - 1], dpb[i - 2] + nums[i - 1]);
        }
        return Math.max(dpa[n], dpb[n]);
    }
}

File: src/Python/213.HouseRobberII.py

class Solution:
    def rob(self, nums):
        n = len(nums)
        if n == 1:
            return nums[0]
        dpa = [0] * (n + 1)
        dpb = [0] * (n + 1)
        for i in range(2, n + 1):
            dpa[i] = max(dpa[i - 1], dpa[i - 2] + nums[i - 2])
            dpb[i] = max(dpb[i - 1], dpb[i - 2] + nums[i - 1])
        return max(dpa[n], dpb[n])


if __name__ == "__main__":
    solution = Solution()
    print(solution.rob([2,3,2]))
    print(solution.rob([1,2,3,1]))
